File: Labs/Lab06/lab06.cpp
#include <iostream>
#include <string>
#include <unordered_map>

std::unordered_map<int, int> mod_inverses = {
    {5, 29}, {7, 31}, {11, 23}, {13, 25},
    {17, 17}, {19, 19}, {23, 11}, {25, 13},
    {29, 5}, {31, 7}, {35, 35}
};

namespace Ciphers {
    const std::string ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    std::string caesar_cipher(const std::string& text, int shift) {
        std::string result;
        for (char c : text) {
            if (std::isalpha(c)) {
                char base = std::isupper(c) ? 'A' : 'a';
                result += (c - base + shift + 26) % 26 + base;
            } else {
                result += c;
            }
        }
        return result;
    }

    std::string affine_cipher(const std::string& text, int a, int b) {
        std::string result;
        int mod = ALPHABET.size();
        
 
        if (mod_inverses.find(a) == mod_inverses.end()) {
            throw std::invalid_argument("a and mod have to be coprime");
        }

      
        int mod_inv = mod_inverses[a];

        for (char c : text) {
            if (std::isalpha(c)) {
                char base = std::isupper(c) ? 'A' : 'a';
                int index = (mod_inv * (c - base - b + mod)) % mod;
                result += index + base;
            } else {
                result += c;
            }
        }
        return result;
    }
}

int main() {
    std::string text;
    std::cout << "Enter a text: ";
    std::getline(std::cin, text);

    std::string caesar_encrypted = Ciphers::caesar_cipher(text, 3);
    std::string caesar_decrypted = Ciphers::caesar_cipher(caesar_encrypted, -3);

   
    std::string affine_encrypted = Ciphers::affine_cipher(text, 5, 8); // Using 5 as 'a'
    std::string affine_decrypted = Ciphers::affine_cipher(affine_encrypted, 5, 8); // Using 5 as 'a'

  
    std::cout << "Caesar Cipher:\n";
    std::cout << "  Encrypted: " << caesar_encrypted << std::endl;
    std::cout << "  Decrypted: " << caesar_decrypted << std::endl;

    std::cout << "Affine Cipher:\n";
    std::cout << "  Encrypted: " << affine_encrypted << std::endl;
    std::cout << "  Decrypted: " << affine_decrypted << std::endl;

    return 0;
}

File: Labs/Lab06/lab06.rb
class Cipher

  
  def initialize(alphabet)
    @alphabet = alphabet
    @multiplicative_inverses = {
      5 => 29, 7 => 31, 11 => 23, 13 => 25, 17 => 17,
      19 => 19, 23 => 11, 25 => 13, 29 => 5, 31 => 7, 35 => 35
    }
  end

  def caesar_cipher(text, shift)
    text.chars.map do |char|
      if @alphabet.include?(char.upcase)
        index = (@alphabet.index(char.upcase) + shift) % @alphabet.size
        @alphabet[index]
      else
        char
      end
    end.join
  end

  def affine_cipher(text, a, b)
    mod = @alphabet.size
    mod_inv = @multiplicative_inverses[a] || raise(ArgumentError, "a and mod must be coprime")

    text.chars.map do |char|
      if @alphabet.include?(char.upcase)
        index = (mod_inv * (@alphabet.index(char.upcase) - b)) % mod
        @alphabet[index]
      else
        char
      end
    end.join
  end
end

cipher = Cipher.new('ABCDEFGHIJKLMNOPQRSTUVWXYZ')

plaintext = "HELLO WORLD"
encrypted_caesar = cipher.caesar_cipher(plaintext, 3)
decrypted_caesar = cipher.caesar_cipher(encrypted_caesar, -3)

encrypted_affine = cipher.affine_cipher(plaintext, 5, 8)
decrypted_affine = cipher.affine_cipher(encrypted_affine, 5, 8)

puts "Caesar Cipher:"
puts "  Encrypted: #{encrypted_caesar}"
puts "  Decrypted: #{decrypted_caesar}"

puts "Affine Cipher:"
puts "  Encrypted: #{encrypted_affine}"
puts "  Decrypted: #{decrypted_affine}"
